### ISSUE TITLE
[Phase 2] Dimensionality Reduction Pipeline

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/high-dim-anomaly-detection#3]
## Summary

Implement dimensionality reduction using Kernel PCA and Autoencoder (MLPRegressor). This phase creates the feature compression layer that makes high-dimensional data tractable for anomaly detection.

**Part of:** Meta Issue #1
**Dependencies:** Phase 1 (base classes, config structure)

## Objectives

1. Implement Kernel PCA wrapper for non-linear reduction
2. Build autoencoder using sklearn MLPRegressor
3. Create reduction pipeline combining methods
4. Extract bottleneck activations for reduced features

## Files to Create

| File | Lines (est.) | Action | Description |
|------|-------------|--------|-------------|
| `src/reduction/__init__.py` | 1 | Create | Package marker |
| `src/reduction/base_reducer.py` | 50 | Create | Abstract reducer interface |
| `src/reduction/kernel_pca.py` | 80 | Create | Kernel PCA wrapper |
| `src/reduction/autoencoder.py` | 150 | Create | MLPRegressor autoencoder |
| `src/reduction/pipeline.py` | 100 | Create | Combine reducers |
| `tests/test_reduction.py` | 120 | Create | Reduction tests |

### Base Reducer
- [ ] Create `BaseReducer` abstract class
- [ ] Define `fit()`, `transform()`, `fit_transform()` interface
- [ ] Add `reconstruct()` method for reconstruction-based scoring
- [ ] Add `reconstruction_error()` method

### Kernel PCA
- [ ] Implement `KernelPCAReducer` wrapping sklearn KernelPCA
- [ ] Configure kernel type, gamma, n_components via YAML
- [ ] Handle memory issues with `eigen_solver='arpack'`
- [ ] Implement approximate reconstruction using `inverse_transform()`

### Autoencoder
- [ ] Implement `AutoencoderReducer` using MLPRegressor
- [ ] Architecture: input -> hidden -> bottleneck -> hidden -> output
- [ ] Extract bottleneck activations via `_forward_to_bottleneck()`
- [ ] Configure hidden_size, n_components, max_iter via YAML
- [ ] Add early stopping for training stability

### Reduction Pipeline
- [ ] Create `ReductionPipeline` combining reducers
- [ ] Support sequential application (KernelPCA -> then autoencoder)
- [ ] Support parallel application (both, then combine)
- [ ] Add caching for fitted reducers

## Code Snippet: BaseReducer

```python
# src/reduction/base_reducer.py
"""Base class for dimensionality reducers."""

from abc import ABC, abstractmethod
import numpy as np


class BaseReducer(ABC):
    """Abstract base class for all reducers."""

    def __init__(self, config: dict):
        self.config = config
        self.is_fitted = False

    @abstractmethod
    def fit(self, X: np.ndarray) -> 'BaseReducer':
        """Fit reducer to data."""
        pass

    @abstractmethod
    def transform(self, X: np.ndarray) -> np.ndarray:
        """Transform data to reduced representation."""
        pass

    def fit_transform(self, X: np.ndarray) -> np.ndarray:
        """Fit and transform in one call."""
        self.fit(X)
        return self.transform(X)

    def reconstruct(self, X: np.ndarray) -> np.ndarray:
        """Reconstruct original from reduced (if supported)."""
        raise NotImplementedError("Reconstruction not supported")

    def reconstruction_error(self, X: np.ndarray) -> np.ndarray:
        """Compute per-sample reconstruction error."""
        X_recon = self.reconstruct(X)
        return np.mean((X - X_recon) ** 2, axis=1)
```

## Code Snippet: AutoencoderReducer

```python
# src/reduction/autoencoder.py
"""Autoencoder using sklearn MLPRegressor."""

import numpy as np
from sklearn.neural_network import MLPRegressor
from sklearn.preprocessing import StandardScaler
from .base_reducer import BaseReducer


class AutoencoderReducer(BaseReducer):
    """Autoencoder for dimensionality reduction."""

    def __init__(self, config: dict):
        super().__init__(config)
        ae_config = config.get('reduction', {}).get('autoencoder', {})
        self.n_components = ae_config.get('n_components', 10)
        self.hidden_size = ae_config.get('hidden_size', 64)
        self.max_iter = ae_config.get('max_iter', 500)
        self.scaler = StandardScaler()
        self.encoder = None

    def fit(self, X: np.ndarray) -> 'AutoencoderReducer':
        X_scaled = self.scaler.fit_transform(X)

        # Architecture: [hidden, bottleneck, hidden]
        self.encoder = MLPRegressor(
            hidden_layer_sizes=(self.hidden_size, self.n_components, self.hidden_size),
            activation='relu',
            solver='adam',
            max_iter=self.max_iter,
            random_state=42,
            early_stopping=True,
            validation_fraction=0.1
        )

        # Train to reconstruct input
        self.encoder.fit(X_scaled, X_scaled)
        self.is_fitted = True
        return self

    def transform(self, X: np.ndarray) -> np.ndarray:
        if not self.is_fitted:
            raise RuntimeError("Reducer not fitted")
        X_scaled = self.scaler.transform(X)
        return self._forward_to_bottleneck(X_scaled)

    def _forward_to_bottleneck(self, X: np.ndarray) -> np.ndarray:
        """Extract bottleneck layer activations."""
        coefs = self.encoder.coefs_
        intercepts = self.encoder.intercepts_

        # First hidden layer (ReLU)
        h1 = np.maximum(0, X @ coefs[0] + intercepts[0])
        # Bottleneck layer (ReLU)
        bottleneck = np.maximum(0, h1 @ coefs[1] + intercepts[1])
        return bottleneck

    def reconstruct(self, X: np.ndarray) -> np.ndarray:
        X_scaled = self.scaler.transform(X)
        X_recon = self.encoder.predict(X_scaled)
        return self.scaler.inverse_transform(X_recon)
```

## Technical Challenges

| Challenge | Solution | Fallback |
|-----------|----------|----------|
| Kernel PCA memory O(n^2) | Use `n_components` wisely, sample for kernel matrix | Switch to standard PCA |
| MLPRegressor doesn't converge | Adjust `max_iter`, `learning_rate_init`, enable `early_stopping` | Use PCA reconstruction error |
| Bottleneck extraction requires weight access | Manual forward pass through `coefs_` | Return reconstruction error instead |

## Verification

```python
# Test reduction pipeline
from src.reduction.autoencoder import AutoencoderReducer
import numpy as np

X = np.random.randn(500, 50)  # 50 features

config = {
    'reduction': {
        'autoencoder': {
            'n_components': 10,
            'hidden_size': 32,
            'max_iter': 200
        }
    }
}

reducer = AutoencoderReducer(config)
reducer.fit(X)

X_reduced = reducer.transform(X)
print(f"Reduced: {X.shape} -> {X_reduced.shape}")  # (500, 50) -> (500, 10)

error = reducer.reconstruction_error(X)
print(f"Reconstruction error: {error.mean():.4f}")
```

## Definition of Done

- [ ] Kernel PCA reduces dimensionality measurably
- [ ] Autoencoder trains without NaN/divergence
- [ ] Reconstruction error < 0.5 (normalized data)
- [ ] Bottleneck activations extractable
- [ ] Pipeline chains reducers correctly
- [ ] Unit tests pass

## Learning Notes

**Concepts to understand:**
- Principal Component Analysis (PCA) and kernels
- Autoencoder architecture (encoder-bottleneck-decoder)
- sklearn MLPRegressor internals (`coefs_`, `intercepts_`)
- Reconstruction error as anomaly signal
